pkg/nsx/services/expand.go

```diff
@@ -0,0 +1,308 @@
+
+	podSelector, namespaces, err := service.getPodSelector(obj, rule)
+	if err != nil {
+		return nil, err
+	}
+	if len(namespaces) == 0 {
+		namespaces = append(namespaces, obj.Namespace)
+	}
+
+	for _, namespace := range namespaces {
+		podsList := &v1.PodList{}
+		podSelector.Namespace = namespace
+		log.V(2).Info("port", "podSelector", podSelector)
+		err := service.Client.List(context.Background(), podsList, podSelector)
+		if err != nil {
+			return nil, err
+		}
+		for _, pod := range podsList.Items {
+			addr, err := service.resolvePodPort(pod, &spPort)
+			if errors.As(err, &util2.PodIPNotFound{}) || errors.As(err, &util2.PodNotRunning{}) {
+				return nil, err
+			}
+			address = append(address, addr...)
+		}
+	}
+
+	if len(address) == 0 {
+		return nil, util2.NoFilteredPod{
+			Desc: "no pod has the corresponding named port, please check the pod selector or labels of CR",
+		}
+	}
+	return util.MergeAddressByPort(address), nil
+}
+
+// Check port name and protocol, only when the pod is really running, and it does have effective ip.
+func (service *SecurityPolicyService) resolvePodPort(
+	pod v1.Pod,
+	spPort *v1alpha1.SecurityPolicyPort,
+) (
+	[]util.PortAddress, error,
+) {
+	var addr []util.PortAddress
+	for _, container := range pod.Spec.Containers {
+		for _, port := range container.Ports {
+			log.V(2).Info("resolvePodPort", "namespace", pod.Namespace, "pod_name", pod.Name,
+				"port_name", port.Name, "containerPort", port.ContainerPort,
+				"protocol", port.Protocol, "podIP", pod.Status.PodIP)
+			if port.Name == spPort.Port.String() && port.Protocol == spPort.Protocol {
+				if pod.Status.Phase != "Running" {
+					errMsg := fmt.Sprintf("pod %s/%s is not running", pod.Namespace, pod.Name)
+					return nil, util2.PodNotRunning{Desc: errMsg}
+				}
+				if pod.Status.PodIP == "" {
+					errMsg := fmt.Sprintf("pod %s/%s ip not initialized", pod.Namespace, pod.Name)
+					return nil, util2.PodIPNotFound{Desc: errMsg}
+				}
+				addr = append(
+					addr,
+					util.PortAddress{Port: int(port.ContainerPort), IPs: []string{pod.Status.PodIP}},
+				)
+			}
+		}
+	}
+	return addr, nil
+}
+
+// Build an ip set group for NSX.
+func (service *SecurityPolicyService) buildRuleIPSetGroup(
+	obj *model.Rule,
+	ips []string,
+) *model.Group {
+	ipSetGroup := model.Group{}
+
+	ipSetGroupID := fmt.Sprintf("%s_ipset", *obj.Id)
+	ipSetGroup.Id = &ipSetGroupID
+	ipSetGroupName := fmt.Sprintf("%s-ipset", *obj.DisplayName)
+	ipSetGroup.DisplayName = &ipSetGroupName
+
+	addresses := data.NewListValue()
+	for _, ip := range ips {
+		addresses.Add(data.NewStringValue(ip))
+	}
+
+	blockExpression := data.NewStructValue(
+		"",
+		map[string]data.DataValue{
+			"resource_type": data.NewStringValue("IPAddressExpression"),
+			"ip_addresses":  addresses,
+		},
+	)
+	ipSetGroup.Expression = append(ipSetGroup.Expression, blockExpression)
+	return &ipSetGroup
+}
+
+// Different direction rule decides different target of the traffic, we should carefully get
```

Comment:
maybe emphasize that named port only care destination selector. that means IN, rule/policy appliedTo, OUT rule's destination.